### DISCUSSION TITLE
RfC - make registration of placeholders possible independently from Templates - resolved ✅

### DISCUSSION THREAD

**Discussion by macolo** [django-cms/django-cms#7927] [General]
## Intro

- django CMS discovers placeholders by scanning (HTML / django) Templates.
- django CMS uses this information to populate the structure sidebar (i.e. which placeholders can be accessed and edited in the sidebar on which page)
- This works generally ok however there are some use cases where a more solid approach would be beneficial:
   - placeholders in apphook applications' templates
   - for #7835 

## Possible solutions
- a global `settings.py` configuration of placeholders (disadvantage: removing or adding a placeholder requires changes in two places, 1) the template and 2) in `settings.py`)
- TBD
- ...

## Solution Approach
- this RfC is being addressed in https://github.com/django-cms/django-cms/pull/7850

**Comment by fsbraun** [comment]
I'd suggest a hierarchical approach:

1. If templates are defined, through templates (`CMS_TEMPLATES` setting)
   * Backwards compatible
   * Only one place to change if using both django CMS' head **and** API
   * Allows for additional config, namely inheritance. 
 2. If no templates are defined, fall back to a new setting, say, 
    ```
    CMS_PLACEHOLDERS = (
        ("home", ("content", "sidebar"), _("Home page")),  # Give slot names
        ("products", ("intro", "ads", "disclaimer"), _("Product page")),
    )
    ```
3. If no setting exists at all: Just one placeholder per page, called `"content"`.

If inheritance is important for a headless solution, the `CMS_PLACEHOLDERS` could add an optional config, despite the fact it might look clunky:
```
CMS_PLACEHOLDERS = (
    ("home", (
        "content",  # Shorthand for {"slot": "content", "inherit": False}
        {
            "slot": "sidebar",
            "inherit": True,
        }
    ), _("Home page")),
    ...
)
```

See here for an early test run: https://github.com/django-cms/django-cms/pull/7850#issuecomment-2125923041

**Comment by fsbraun** [comment]
On another note, allowing for an alternative method of placeholder declaration - besides within templates - will simplify certain use cases for app hooks:
* Currently, app hooks need to define a `get_template()` method on their frontend-editable models to provide a template for placeholder scanning only. Often, it only lists the placeholders using the`{% placeholder %}` template tag.
* Typically, a different template is used for actually displaying any model content using `{% render_placeholder %}` instead of `{% placeholder %}`

I suggest to also checking models for a `get_placeholder_slots()` method which will return a model's used slots. It will be evaluated in cases the `get_template()` method does not exist or returns a valid template.

**Reply by mrbazzan** [reply]
I think for apphooks, even if `get_placeholder_slots()` is defined;
there is still a need to define `get_template()`.

Is it possible not to have to define `get_template()`?

**Reply by fsbraun** [reply]
Yes, for headless apps or apps that are edited in the structure mode only (full-screen structure board).
